rfc/0002/README.md

```diff
@@ -0,0 +1,165 @@
+Rabbit storage for containerized applications
+=============================================
+
+**authors**: Nate Roiger <nate.roiger@hpe.com> |
+**state**: ideation
+
+---
+
+For Rabbit to provide storage to a containerized application there needs to be _some_ mechanism. The remainder of this RFC proposes that mechanism.
+
+Actors
+------
+
+There are several different actors involved
+
+- The AUTHOR of the containerized application
+- The ADMINISTRATOR who works with the author to determine the application requirements for execution
+- The USER who intends to to use the application using the 'container' directive in their job specification
+- The RABBIT software that interprets the #DWs and starts the container during execution of the job
+
+There are multiple relationships between the actors:
+- AUTHOR to ADMINISTRATOR: The author defines how their application is executed and the NNF storage requirements
+- Between the AUTHOR and USER: The application expects certain storage, and the #DW must meet those expectations.
+- ADMINISTRATOR to RABBIT: Admin tells Rabbit how to run the container
+- Between USER and RABBIT: User provides the #DW container directive in the job specification, and Rabbit validates & interprets the directive.
+
+Proposal
+--------
+
+The proposal below might take a couple of read-throughs; I've also added a concrete example afterward that might help.
+
+1. The AUTHOR writes their application expecting NNF Storage at specific locations. For each storage requirement, they define...
+    1. a unique name for the storage which can be referenced in the 'container' directive
+    2. the expected storage types; if necessary
+    3. the required mount path or mount path prefix
+    4. other constraints or storage requirements (i.e. minimum capacity)
+2. The AUTHOR works with the ADMINISTRATOR to define, a unique name for the program, the pod specification for executing their program, and the NNF storage requirements described above. The ADMINISTRATOR creates a corresponding _NNF Container Profile_ custom kubernetes resource.
+3. The USER who desires to use the application works with the AUTHOR and the related Container Profile to understand the storage requirements.
+4. The USER submits a WLM job with the #DW container fields populated
+5. WLM runs the job and drives the job through the following stages...
+    1. Proposal: RABBIT validates the #DW container directive by comparing the supplied values to what is listed in the Container Profile. If the USER fails to meet the requirements, the job fails.
+    2. Pre-run: RABBIT software...
+        1. creates a config map reflecting the storage requirements and any runtime parameters
+        2. duplicates the pod specification from the Container Profile and patches the necessary Volumes and the config map. The spec is used as the basis for starting the necessary pods and containers.
```

Comment:
The pod spec will be used in a deployment/daemonset, to make the container run on the rabbit nodes that go with the Flux job.  Rabbit software will add labels and tolerations to the deployment/daemonset to let k8s schedule this pod on the rabbits.

Reply:
The end result is pods run on the Rabbits, It doesn't matter how they get there. I don't think we need to mention deployment/daemonset since that just complicates the document.